src/csharp/Grpc.Core/Internal/CompletionQueueSafeHandle.cs

```diff
@@ -29,7 +29,7 @@ internal class CompletionQueueSafeHandle : SafeHandleZeroIsInvalid
     {
         static readonly NativeMethods Native = NativeMethods.Get();
 
-        AtomicCounter shutdownRefcount = new AtomicCounter(1);
+        long shutdownRefcount = AtomicCounter.Create(1);
```

Comment:
The async completion queues (used by the threadpool) are only initialized when GrpcEnvironment (and the internal threadpool) is started, so this overhead is negligible compared to the costs of initializing the grpc stack.

On the other hand, if you're using sync unary calls, a new completion queue is created for each call (that's how the C-core native API works) - I think in this case you're onto something and we actually might be initializing one AtomicCounter per RPC. Let me look into that.

https://github.com/grpc/grpc/blob/4a2f2dd1f1a65a8463e4002d6edaf9ddcbd9db7a/src/csharp/Grpc.Core/Internal/AsyncCall.cs#L81

All the other allocations of AtomicCounter in this pull request are in situations that happen too rarely to actually matter IMHO.

Reply:
After some inspection, it looks like the shutdownRefCount is only useful for completion queues that are created using CreateAsync() (the ones used in the internal threadpool, not the ones allocated when sync unary call is invoked).
@mgravell  Could you check that all the per-call allocations you're seeing in the memory profile are coming from the sync unary call case mentioned here?

Reply:
@jtattermusch yes, the reason it was showing on my profiler was because of the sync unary; by the time I'd fixed it there, it made sense to be consistent and apply the same fix everywhere

Reply:
I think in that case we should just avoid using the counter for sync unary call completely (AFAIK it shouldn't be necessary). I quite like the encapsulation of AtomicCounter and it seems that the refactoring provides little performance benefit if we eliminate the sync unary call case.

Reply:
that would work too - if it can't be cleanly excised, just leave the reference `null` and make the access `null`-safe?